File: sources/lua/14_file_read/main.lua
-- task 14 file_read — expected output: 484442112
-- build: none (interpreted)    run: lua main.lua (PUC Lua 5.4)    also runs on LuaJIT: luajit main.lua
-- data.bin (100 MiB: the bytes 0..255 repeating) must sit in the working directory.
-- Read in 1 MiB chunks, then scan every byte.

local f = io.open("data.bin", "rb")

local total = 0
while true do
    local chunk = f:read(1048576)
    if not chunk then
        break
    end
    for i = 1, #chunk do
        total = total + string.byte(chunk, i)
    end
end

f:close()
print(total % 4294967296)

File: sources/r/14_file_read/main.R
# task 14 file_read — expected output: 484442112
# build: Rscript main.R    run: Rscript main.R
#
# data.bin is 104857600 bytes, the bytes 0..255 repeating, opened from the
# working directory and read in 1 MiB chunks, never a byte per syscall. R has
# no per-byte loop worth running, so each chunk is summed vectorised; the
# accumulation is exact because the total before the modulus (13369344000)
# is far below 2^53. The modulus is taken once, at the end.

read_file <- function() {
  con <- file("data.bin", "rb")
  total <- 0
  repeat {
    chunk <- readBin(con, "raw", n = 1048576)
    if (length(chunk) == 0) break
    total <- total + sum(as.integer(chunk))
    if (length(chunk) < 1048576) break
  }
  close(con)
  cat(sprintf("%.0f\n", total %% 4294967296))
}

read_file()
